frontend/lib/resolution/split-init.cpp

```diff
@@ -227,196 +235,159 @@ void FindSplitInits::handleYield(const uast::Yield* ast, RV& rv) {
-            x = 5;
-            compilerWarning('in if branch, x.type is ' + x.type:string);
-          } else {
-            x = 12.0;
-            compilerWarning('in else branch, x.type is ' + x.type:string);
-          }
-        }
 
-        If it succeeds at compiling, the type of 'x' will seem to change
-        as it compiles. It's OK to be temporarily wrong about a type
-        like this if we are going to issue an error, but to do so
-        for correct code seems problematic.
-     */
-    for (size_t i = 0; i < size; i++) {
-      QualifiedType thenRhsType = splitInitedBothThenTypes[i];
-      QualifiedType elseRhsType = splitInitedBothElseTypes[i];
-      if (thenRhsType != elseRhsType) {
-        context->error(cond,
-                       "types do not match in conditional split init");
-      }
+  // calculate the set of variables that are initialized in all 
+  // frames that do not return or throw
+  std::set<ID> locallySplitInitedVars;
+  // calculate the set of variables that are initialized in all 
+  // frames. 
+  std::set<ID> splitInitedInAll;
+  for (auto id : locallyInitedVars) {
+    bool allFramesInit = true;
+    bool allFramesInitReturnOrThrow = true;
+    for(auto frame : frames) {
+      bool thisFrameInits = frame->initedVars.count(id) > 0;
+      allFramesInit &= thisFrameInits;
+      allFramesInitReturnOrThrow &= (thisFrameInits || frame->returnsOrThrows);
+    }
+    if (allFramesInit) {
+      splitInitedInAll.insert(id);
+    }
+    if (allFramesInitReturnOrThrow) {
+      locallySplitInitedVars.insert(id);
+    } else {
+      currentFrame->mentionedVars.insert(id);
     }
   }
 
-  // propagate the mentioned variables
-  if (thenFrame != nullptr) {
-    for (const auto& id : thenFrame->mentionedVars) {
-      if (thenFrame->declaredVars.count(id) == 0) {
-        frame->mentionedVars.insert(id);
-      }
+  if (!total) {
+    for (auto frame: frames) {
+      propogateMentions(currentFrame, frame);
+    } 
+    for (const auto & id: locallySplitInitedVars) {
+      currentFrame->mentionedVars.insert(id);
     }
+    return;
   }
-  if (elseFrame != nullptr) {
-    for (const auto& id : elseFrame->mentionedVars) {
-      if (elseFrame->declaredVars.count(id) == 0) {
-        frame->mentionedVars.insert(id);
+
+  // propogate the initialization points for split init variables
+  // also gather the split-inited variable order and types for errory checking
+  std::vector<std::vector<ID>> splitInitedAllIds;
+  std::vector<std::vector<QualifiedType>> splitInitedAllTypes;
+  for (auto frame : frames) {
+    std::vector<ID> currIds;
+    std::vector<QualifiedType> currTypes;
+    for (auto pair : frame->initedVarsVec) {
+      ID id = pair.first;
+      QualifiedType rhsType = pair.second;
+      if (locallySplitInitedVars.count(id) > 0) {
+        addInit(currentFrame, id, rhsType);
+      }
+      if (splitInitedInAll.count(id) > 0) {
+        currIds.push_back(id);
+        currTypes.push_back(rhsType);
```

Comment:
Skeptical of this. Are you saying that if one branch returns, and two others have swapped initialization order, this won't complain?

```Chapel
var x, y;
select something {
  when one { x = 1; y = 2; }
  when two { return; }
  otherwise { y = 1; x = 2; }
}
```
Here, initialization order constraints are violated, but not checked (?). Am I misunderstanding something? The fix is also not obvious to me since you can't use the first branch as a reference:

```Chapel
var x, y, z;
select something {
  when one { z = 0; return; }
  when two { z = 0; x = 1; y = 2; }
  otherwise { z = 0; y = 1; x = 2; }
}
```

What's really going on? It looks to me like the initialization is valid iff all `initedVarsVec` are prefixes of the same sequence. In this case, the sequence is `x, y ,z`. How do we elegantly express this?

I suggest making `checkForErrors` into a `finalizeVariableOrder` function, and then call `addInit` on the results of that. I don't actually think you need to keep a vector of vectors, since you have the invariant that for `idx1, idx2 < i` in the `checkForErrors` loop, `ids[idx1] == ids[idx2]` and `types[idx1] == types[idx2]` -- they're the same sequence! Some pseudocode-ish logic:

```C++
std::vector<std::pair<ID, QualifiedType>>
FindSplitInits::finalizeVariableOrder(const std::vector<VarFrame*>& frames) {
  std::vector<std::pair<ID, QualifiedType>> referenceDecls;
  
  for (auto frame : frames) {
    size_t idx = 0;
    for (auto& pair : frame->initedVarsVec) {
      auto& id = pair.first;
      auto& qt = pair.second;
     
      // This was an entry corresponding to a frame-local variable initialization.
      if (locallySplitInitedVars.count(id) == 0) continue;

      if (idx > referenceDecls.size()) {
        // Other branches returned before initializing the rest of the variables.
        // We are now defining the 'reference' sequence.
        referenceDecls.emplace_back(id, qt);
      } else {
        // Make sure that this is also the variable that was next initialized in other branches.
        if (referenceDecls[idx].first != id) { /* emit the error here */ return; }
        if (referenceDecls[idx].second != qt) { /* emit the error here */ return; }
      }
      idx++;
  }
  return referenceDecls;
}
```

From your code in `handleDisjunction`, you can iterate over the finalized order and call `addInit`, since those are the guaranteed-correct variables that were initialized in all branches that don't throw.

Reply:
b.t.w., the spec seems to say that `select` statements aren't even considered for split-init (https://chapel-lang.org/docs/1.33/language/spec/variables.html#split-initialization), so I am having trouble checking the production behavior.

Reply:
> Skeptical of this. Are you saying that if one branch returns, and two others have swapped initialization order, this won't complain?

I think you are correct here. The spec gives: "Any variables declared in a particular scope that are initialized with split init in both the `then` and `else` branches of a conditional must be initialized in the same order in the `then` and `else` branches." The code implements "Any variable declared in a particular scope that are initialized with split init in all branches of a conditional or `select` statement must be initialized in the same order in all branches."

Generalizing the spec could also arrive at something like  "Any variables declared in a particular scope that are initialized with split init in multiple branches of a conditional or `select` statement must be initialized in the same order in all branches in which those variables are split init." But here's another case:
```
var x,y,z;
select something3 {
  when 1 { z = 0; y = 1; return;}
  when 2 {y = 2; x = 0; return;}
  otherwise {z = 1; x = 4;}
}
```
All variables split-initialized in each branch are initialized in the order `z y x`, but not all variables are present in each branch. This seems like a pain to implement.

I think your approach resolves some of these issues, but I have some reservations. First, how to formulate the requirement for the spec. "Variables declared in a particular scope that are initialized with split init in multiple branches of a conditional or `select` statement must be initialized in orders that are prefixes of the same sequence" is somewhat confusing I think. Second, it seems somewhat arbitrary that we require prefix orders, aside from the complexity of implementing the logic to make the check.

One thing that I think would help me get on board is a better understanding of the purpose of the initialization order requirement. Do you have any insight there?

Reply:
Also, I've got another PR draft for updating the spec for these cases, so once we figure something out i'll add it there too!